lib/data_cite/client.rb

```diff
@@ -0,0 +1,126 @@
+# frozen_string_literal: true
+
+require 'faraday'
+
+module DataCite
+  class Client
+    class Error < StandardError; end
+
+    def initialize(prefix: nil, publisher: nil)
+      @prefix = prefix
+      @publisher = publisher
+    end
+
+    def register(suffix = nil)
+      attributes = if suffix.present?
+                     { doi: "#{prefix}/#{suffix}" }
+                   else
+                     { prefix: prefix }
+                   end
+
+      data = {
+        type: 'dois',
+        attributes: attributes
+      }
+
+      process_response post(data: data)
+    end
+
+    def publish(doi: nil, metadata:)
+      data = {
+        type: 'dois',
+        attributes: metadata.merge(
+          event: 'publish',
+          publisher: publisher
+        )
+      }
+
+      if doi.blank?
+        data[:attributes][:prefix] = prefix
+        process_response post(data: data)
+      else
+        data[:id] = doi
+        process_response put(doi: doi, data: data)
+      end
+    end
+
+    def update(doi:, metadata:)
+      data = {
+        type: 'dois',
+        id: doi,
+        attributes: metadata
+      }
+
+      process_response put(doi: doi, data: data)
+    end
+
+    def delete(doi:)
+      process_response connection.delete("/dois/#{doi}")
+    end
+
+    def prefix
+      @prefix ||= ENV['DATACITE_PREFIX']
+    end
+
+    def publisher
+      @publisher ||= ENV['DATACITE_PUBLISHER']
+    end
+
+    private
+
+      def post(data:)
+        connection.post do |req|
+          req.headers['Content-Type'] = 'application/vnd.api+json'
+          req.body = { data: data }.to_json
+        end
+      end
+
+      def put(doi:, data:)
+        connection.put("/dois/#{doi}") do |req|
+          req.headers['Content-Type'] = 'application/vnd.api+json'
+          req.body = { data: data }.to_json
+        end
+      end
+
+      def process_response(response)
+        parsed_body = begin
+                        JSON.parse(response.body)
+                      rescue JSON::ParserError
+                        {}
+                      end
+
+        # @todo Consider raising more specific errors, i.e.
+        # MethodNotAllowedError under certain cases
+        unless response.success?
+          errors = parsed_body.fetch('errors', [])
+          stringified_errors = errors.inspect
+          message = stringified_errors.presence || "DataCite response returned #{response.status}"
+
+          raise Error.new(message)
+        end
+
+        doi = parsed_body.dig('data', 'id')
+
+        [doi, parsed_body]
```

Comment:
@rschenk According to Code Climate, we don't seem to be testing this method. 🤷‍♂